composer/callbacks/run_directory_uploader.py

```diff
@@ -0,0 +1,323 @@
+                self._object_name_prefix = ""
+            else:
+                if not object_name_prefix.endswith('/'):
+                    object_name_prefix = f"{object_name_prefix}/"
+                self._object_name_prefix = object_name_prefix
+        self._last_upload_timestamp = 0.0  # unix timestamp of last uploaded time
+        if upload_staging_folder is None:
+            self._tempdir = tempfile.TemporaryDirectory()
+            self._upload_staging_folder = self._tempdir.name
+        else:
+            self._tempdir = None
+            self._upload_staging_folder = upload_staging_folder
+
+        if num_concurrent_uploads < 1:
+            raise ValueError("num_concurrent_uploads must be >= 1. Blocking uploads are not supported.")
+        self._num_concurrent_uploads = num_concurrent_uploads
+        self._provider = provider
+        self._container = container
+
+        _validate_credentials(provider, container, self._object_name_prefix, provider_init_kwargs)
+
+        if use_procs:
+            mp_ctx = multiprocessing.get_context('spawn')
+            self._file_upload_queue: Union[queue.Queue[str],
+                                           multiprocessing.JoinableQueue[str]] = mp_ctx.JoinableQueue()
+            self._finished_cls: Union[Callable[[], multiprocessing._EventType], Type[threading.Event]] = mp_ctx.Event
+            self._proc_class = mp_ctx.Process
+        else:
+            self._file_upload_queue = queue.Queue()
+            self._finished_cls = threading.Event
+            self._proc_class = threading.Thread
+        self._finished: Union[None, multiprocessing._EventType, threading.Event] = None
+        self._workers = []
+
+    def init(self, state: State, logger: Logger) -> None:
+        if get_run_directory() is None:
+            return
+        del state, logger  # unused
+        self._finished = self._finished_cls()
+        self._last_upload_timestamp = 0.0
+        self._workers = [
+            self._proc_class(target=_upload_worker,
+                             kwargs={
+                                 "file_queue": self._file_upload_queue,
+                                 "is_finished": self._finished,
+                                 "upload_staging_dir": self._upload_staging_folder,
+                                 "provider_name": self._provider,
+                                 "container_name": self._container,
+                                 "object_name_prefix": self._object_name_prefix,
+                                 "init_kwargs": self._provider_init_kwargs,
+                             }) for _ in range(self._num_concurrent_uploads)
+        ]
+        for worker in self._workers:
+            worker.start()
+
+    def batch_end(self, state: State, logger: Logger) -> None:
+        if get_run_directory() is None:
+            return
+        if (state.batch_idx + 1) % self._upload_every_n_batches == 0:
+            self._trigger_upload(logger, LogLevel.BATCH)
```

Comment:
Uploads happen `_upload_every_n_batches`, not every batch. The performance hit is the time it takes to copy the files that have changed since the last upload. I can try to measure it.

Reply:
Performance implications:
* Negligible if there are very small files to upload (e.g. throughput drops from ~1280 to ~1260, but only sometimes)
* If you have larger artifacts to upload (e.g. 100 MB traces), then throughput drops from ~1280 to ~1040

Note that only new or modified files are uploaded (determined by using the last modified timestamp), and uploads only occur every `self._upload_every_n_batches`. This is equivalent to the wandb artifacts.

If we wanted performance over correctness, we can do non-blocking copies, but then we may be uploading corrupted data. This can be done in a separate PR (and controlled via a hparams flag) if desired. 

Raw data: https://wandb.ai/mosaic-ml/research2-run/runs/vv4atxxz. The drops at epoch 10, 20 and 30 come from the torch profiler saving the trace. The drops at steps 20, 30, and 40 come from the run directory uploader uploading traces. The other spikes every 10 steps are from the uploader uploading log files.